rfd/0155-scoped-webauthn-credentials.md

```diff
@@ -0,0 +1,197 @@
+  repeated bytes allow_credentials = 3 [(gogoproto.jsontag) = "allowCredentials,omitempty"];
+  bool resident_key = 4 [(gogoproto.jsontag) = "residentKey,omitempty"];
+  string user_verification = 5 [(gogoproto.jsontag) = "userVerification,omitempty"];
+
+  // new:
+  // Scope authorized by this webauthn session. Ex: "login".
+  Scope scope = 6 [(gogoproto.jsontag) = "scope,omitempty"];
+  // UsesRemaining is a count of how many more times this session can be used for authentication
+  // before the session expires and is deleted. When set to 0 (default) or 1, the session will
+  // be deleted after the first authentication.
+  int64 uses_remaining = 7;
+}
+
+// Scope is a scope authorized by a webauthn session.
+enum Scope {
+  SCOPE_UNSPECIFIED = 0;
+  SCOPE_LOGIN = 1;
+  SCOPE_PASSWORDLESS_LOGIN = 2;
+  // MFA device management
+  SCOPE_MANAGE_DEVICES = 3;
+  // Account recovery - reset password, devices, etc.
+  SCOPE_RECOVERY = 4;
+  // Used for per-session MFA and moderated session presence checks
+  SCOPE_SESSION = 5;
+  SCOPE_HEADLESS = 6;
+  SCOPE_ADMIN_ACTION = 7;
+}
+
+// authservice.proto
+
+message CreateAuthenticateChallengeRequest {
+  oneof Request {
+    UserCredentials UserCredentials = 1 [(gogoproto.jsontag) = "user_credentials,omitempty"];
+    string RecoveryStartTokenID = 2 [(gogoproto.jsontag) = "recovery_start_token_id,omitempty"];
+    ContextUser ContextUser = 3 [(gogoproto.jsontag) = "context_user,omitempty"];
+    Passwordless Passwordless = 4 [(gogoproto.jsontag) = "passwordless,omitempty"];
+  }
+  IsMFARequiredRequest MFARequiredCheck = 5 [(gogoproto.jsontag) = "mfa_required_check,omitempty"];
+  
+  // new:
+  // Scope is the authorization scope for this MFA challenge. Ex: login.
+  // Only applies to webauthn challenges.
+  webauthn.Scope Scope = 6 [(gogoproto.jsontag) = "scope,omitempty"];
+  // Uses is the number of times resulting webauthn credentials can be used.
+  // Only applies to webauthn challenges with scopes that allow reuse.
+  int64 Uses = 7 [(gogoproto.jsontag) = "uses,omitempty"];
+}
+```
+
+### Client changes
+
+Clients will be expected to provide a scope when requesting an MFA challenge
+through `rpc CreateAuthenticateChallenge`. For specific login and device
+management endpoints, the scope will be automatically set on the server side.
+
+Clients can optionally provide a number of uses which will be allowed, though
+reuse is only allowed for the admin action scope.
+
+### Server changes
+
+#### Scope
+
+When verifying a webauthn credential against the user's stored webauthn
+challenge, the Auth Server will check that the stored scope matches the scope
+that the Auth server is verifying for. For example, if the Auth server is
+verifying a webauthn credential for scope "login", but webauthn challenge
+stored for the user has scope "headless", the verification will fail.
+
+#### Reuse
+
+After verifying a webauthn credential, the Auth server will check the `Uses`
```

Comment:
Should we enforce a max number of reuses per scope type?

Reply:
I don't think this is necessary for now, it would have to be high anyways. The "cloud invite" button performs 3 admin actions for every user you invite, so the number of reuses needed may be very high.

Reply:
That is a bit scary.